solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/SolrFeature.java

```diff
@@ -120,67 +123,66 @@ protected void validate() throws FeatureException {
       super(SolrFeature.this, searcher, request, originalQuery, efi);
       try {
-        String solrQuery = q;
-        final List<String> fqs = fq;
-
-        if ((solrQuery == null) || solrQuery.isEmpty()) {
-          solrQuery = "*:*";
-        }
-
-        solrQuery = macroExpander.expand(solrQuery);
-        if (solrQuery == null) {
-          throw new FeatureException(this.getClass().getSimpleName()+" requires efi parameter that was not passed in request.");
-        }
-
-        final SolrQueryRequest req = makeRequest(request.getCore(), solrQuery,
-            fqs, df);
+        final SolrQueryRequest req = makeRequest(request.getCore(), q, fq, df);
         if (req == null) {
           throw new IOException("ERROR: No parameters provided");
         }
 
+        // Build the scoring query
+        Query scoreQuery;
+        String qStr = q;
+        if (qStr == null || qStr.isEmpty()) {
+          scoreQuery = null; // ultimately behaves like MatchAllDocs
+        } else {
+          qStr = macroExpander.expand(qStr);
+          if (qStr == null) {
+            throw new FeatureException(this.getClass().getSimpleName() + " requires efi parameter that was not passed in request.");
+          }
+          scoreQuery = QParser.getParser(qStr, req).getQuery();
+          // note: QParser can return a null Query sometimes, such as if the query is a stopword or just symbols
+          if (scoreQuery == null) {
+            scoreQuery = new MatchNoDocsQuery(); // debatable; all or none?
+          }
+        }
+
         // Build the filter queries
-        queryAndFilters = new ArrayList<Query>(); // If there are no fqs we just want an empty list
-        if (fqs != null) {
-          for (String fq : fqs) {
-            if ((fq != null) && (fq.trim().length() != 0)) {
-              fq = macroExpander.expand(fq);
-              if (fq == null) {
-                throw new FeatureException(this.getClass().getSimpleName()+" requires efi parameter that was not passed in request.");
+        Query filterDocSetQuery = null;
+        if (fq != null) {
+          List<Query> filterQueries = new ArrayList<>(); // If there are no fqs we just want an empty list
+          for (String fqStr : fq) {
+            if (fqStr != null) {
+              fqStr = macroExpander.expand(fqStr);
+              if (fqStr == null) {
+                throw new FeatureException(this.getClass().getSimpleName() + " requires efi parameter that was not passed in request.");
               }
-              final QParser fqp = QParser.getParser(fq, req);
-              final Query filterQuery = fqp.getQuery();
+              final Query filterQuery = QParser.getParser(fqStr, req).getQuery();
               if (filterQuery != null) {
-                queryAndFilters.add(filterQuery);
+                filterQueries.add(filterQuery);
               }
             }
           }
+
+          DocSet filtersDocSet = searcher.getDocSet(filterQueries); // execute
```

Comment:
observation: i've sort of convinced myself that at this point `filterQueries` should be a non-empty list but if it were an empty list then i haven't yet dived fully enough into the `getDocSet` implementation to understand if that is doing what we want or if `filterDocSetQuery` remaining `null` might be clearer. perhaps you've thought about this already too?

Reply:
Good eye.  It's possible filterQueries might be empty if say someone added a blank fq.  And I verified in a debugger get getDocSet does the right thing.  It's taking a long time to do its job however (accumulating a complete docset, bit by bit).  In another issue I'm targeting at 9.0, I optimized this case to a live-docs DocSet, among other things relating to TwoPhaseIterator.  I should pull that out and make it its own improvement, along with a bit more documentation to getDocSet to indicate that's what the semantics are if an empty list is passed.